utils/ort/src/main/kotlin/storage/S3FileStorage.kt

```diff
@@ -0,0 +1,175 @@
+import aws.sdk.kotlin.services.s3.model.NotFound
+import aws.sdk.kotlin.services.s3.model.PutObjectRequest
+import aws.sdk.kotlin.services.s3.model.S3Exception
+import aws.smithy.kotlin.runtime.auth.awscredentials.CachedCredentialsProvider
+import aws.smithy.kotlin.runtime.auth.awscredentials.Credentials
+import aws.smithy.kotlin.runtime.content.ByteStream
+import aws.smithy.kotlin.runtime.content.toByteArray
+import aws.smithy.kotlin.runtime.net.Url
+
+import java.io.ByteArrayInputStream
+import java.io.File
+import java.io.InputStream
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.async
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+
+import org.apache.logging.log4j.kotlin.Logging
+
+/**
+ * A [FileStorage] that stores files in an AWS S3 bucket [bucketName]. The [read] and [exists] operations are
+ * blocking, but the [write] operation is asynchronous unless a [customEndpoint] is provided.
+ */
+@Suppress("SwallowedException")
+class S3FileStorage(
+    /**
+     * The name of the S3 bucket used to store files in.
+     */
+    private val bucketName: String,
+    /**
+     * The AWS region to be used.
+     */
+    private val awsRegion: String? = null,
+    /**
+     * The AWS access key.
+     */
+    private val accessKeyId: String? = null,
+    /**
+     * The AWS secret for the access key.
+     */
+    private val secretAccessKey: String? = null,
+    /**
+     * Non-public endpoint for testing.
+     */
+    private val customEndpoint: String? = null
+) : FileStorage {
+    private companion object : Logging
+
+    private val scope = CoroutineScope(Dispatchers.Default)
+
+    val s3 by lazy {
+        val provider = if (accessKeyId != null && secretAccessKey != null) {
+            CachedCredentialsProvider(StaticCredentialsProvider(Credentials(accessKeyId, secretAccessKey)))
+        } else {
+            null
+        }
+
+        val client = if (awsRegion != null && provider != null) {
+            S3Client {
+                region = awsRegion
+                credentialsProvider = provider
+                endpointUrl = if (customEndpoint != null) Url.parse(customEndpoint) else null
+            }
+        } else {
+            runBlocking {
+                if (awsRegion != null) {
+                    S3Client.fromEnvironment {
+                        region = awsRegion
+                    }
+                } else {
+                    S3Client.fromEnvironment()
+                }
+            }
+        }
+
+        client
+    }
+
+    override fun exists(path: String): Boolean {
+        val request = HeadObjectRequest {
+            key = path
+            bucket = bucketName
+        }
+
+        var fileExists = true
+
+        runBlocking {
+            try {
+                s3.headObject(request)
+            } catch (e: NotFound) {
+                fileExists = false
+            }
+        }
```

Comment:
Probably better something like

    return runBlocking { runCatching { s3.headObject(request) }.isSuccess }

Reply:
Thank you for that valuable suggestion. I'd still like to keep some record if the expection is any different form NotFound.